File: frontend/src/features/opencode-go-usage/schemas.ts
import { z } from "zod";

export const OpenCodeGoUsageWindowSchema = z.object({
  window: z.enum(["rolling", "weekly", "monthly"]),
  remainingPercent: z.number().min(0).max(100),
  usedPercent: z.number().min(0).max(100),
  resetsAt: z.string().datetime({ offset: true }),
  capturedAt: z.string().datetime({ offset: true }),
});

export const OpenCodeGoUsageMonitorSchema = z.object({
  configured: z.boolean(),
  lastAttemptAt: z.string().datetime({ offset: true }).nullable(),
  lastSuccessAt: z.string().datetime({ offset: true }).nullable(),
  lastError: z.string().nullable(),
  windows: z.array(OpenCodeGoUsageWindowSchema),
});

export const OpenCodeGoUsageCredentialRequestSchema = z.object({
  apiKey: z.string().trim().min(1),
});

export type OpenCodeGoUsageMonitor = z.infer<typeof OpenCodeGoUsageMonitorSchema>;
export type OpenCodeGoUsageCredentialRequest = z.infer<typeof OpenCodeGoUsageCredentialRequestSchema>;
export type OpenCodeGoUsageWindow = z.infer<typeof OpenCodeGoUsageWindowSchema>;

File: app/modules/opencode_go_usage/schemas.py
from __future__ import annotations

from datetime import datetime

from pydantic import Field

from app.modules.shared.schemas import DashboardModel


class OpenCodeGoUsageWindowResponse(DashboardModel):
    window: str
    remaining_percent: float = Field(ge=0, le=100)
    used_percent: float = Field(ge=0, le=100)
    resets_at: datetime
    captured_at: datetime


class OpenCodeGoUsageMonitorResponse(DashboardModel):
    configured: bool
    last_attempt_at: datetime | None = None
    last_success_at: datetime | None = None
    last_error: str | None = None
    windows: list[OpenCodeGoUsageWindowResponse] = Field(default_factory=list)


class OpenCodeGoUsageCredentialRequest(DashboardModel):
    api_key: str = Field(min_length=1, max_length=4096)
